File: scripts/pack-fabric/routing/lib/legal-topology/find-path-v4.js
"use strict";

/**
 * Turn-aware V4 search. Incoming directed arc is part of visited state.
 * No coincident-node stitches. No proximity repairs.
 */

const {
  compileRestrictionIndex,
  advanceRestrictionState,
  activeKey
} = require("./restrictions");

function haversineMeters(a, b) {
  const toRad = (deg) => (deg * Math.PI) / 180;
  const r = 6371000;
  const dLat = toRad(b[1] - a[1]);
  const dLng = toRad(b[0] - a[0]);
  const lat1 = toRad(a[1]);
  const lat2 = toRad(b[1]);
  const x =
    Math.sin(dLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;
  return 2 * r * Math.atan2(Math.sqrt(x), Math.sqrt(1 - x));
}

function accessCode(pack, ei, from, to) {
  const forward = pack.edgeFrom[ei] === from && pack.edgeTo[ei] === to;
  return pack.edgeAccess[ei * 2 + (forward ? 0 : 1)];
}

function endpointOk(code, isEndpoint, endpointKind = null) {
  if (code === 0) return true;
  if (code === 3) return !!isEndpoint && endpointKind !== "customers";
  if (code === 4) return !!isEndpoint && endpointKind === "customers";
  return false;
}

function findPathV4(pack, geom, origin, dest, options = {}) {
  const { legalSnapDetailed, selectConnectedSnapPair } = require("./snap");
  const allowUnknown = options.allowUnknown === true;
  const startDetailed = legalSnapDetailed(pack, geom, origin, {
    headingDeg: options.startHeadingDeg,
    intentBearingDeg: options.intentBearingDeg,
    maxMeters: options.maxMeters,
    zoom: options.zoom,
    allowUnknown
  });
  const endDetailed = legalSnapDetailed(pack, geom, dest, {
    headingDeg: options.endHeadingDeg,
    intentBearingDeg: options.intentBearingDeg != null ? (options.intentBearingDeg + 180) % 360 : null,
    maxMeters: options.maxMeters,
    zoom: options.zoom,
    allowUnknown
  });
  const picked = selectConnectedSnapPair(
    pack,
    startDetailed.candidates,
    endDetailed.candidates,
    { allowUnknown }
  );
  if (!picked.ok) {
    return {
      ok: false,
      reason: picked.reason || "no_snap",
      unprovenStitches: 0,
      snap: {
        allowUnknown,
        start: startDetailed,
        end: endDetailed,
        rejections: picked.rejections
      }
    };
  }
  const startSnaps = [picked.start];
  const endSnaps = [picked.end];

  const destEdges = new Set(endSnaps.map((s) => s.edgeIndex));
  const restrictionIndex = compileRestrictionIndex(pack.restrictions || []);
  const n = pack.nodeCount;
  const dist = new Map();
  const prev = new Map();
  const heap = [];
  function push(node, incomingEdge, restrictionActive, cost, pathEdges) {
    const key = node + ":" + incomingEdge + ":" + activeKey(restrictionActive);
    const old = dist.get(key);
    if (old != null && old <= cost) return;
    dist.set(key, cost);
    prev.set(key, pathEdges);
    heap.push({ node, incomingEdge, restrictionActive, cost, pathEdges });
  }

  for (const snap of startSnaps) {
    const from = snap.forward ? pack.edgeFrom[snap.edgeIndex] : pack.edgeTo[snap.edgeIndex];
    const to = snap.forward ? pack.edgeTo[snap.edgeIndex] : pack.edgeFrom[snap.edgeIndex];
    const remain = snap.forward
      ? (1 - snap.fraction) * pack.edgeMeters[snap.edgeIndex]
      : snap.fraction * pack.edgeMeters[snap.edgeIndex];
    const code = accessCode(pack, snap.edgeIndex, from, to);
    const isDest = destEdges.has(snap.edgeIndex);
    const startEndpointKind = options.startEndpointKind || null;
    const endpointKind = isDest ? options.endEndpointKind || null : startEndpointKind;
    if (!endpointOk(code, true, endpointKind) && code !== 0) continue;
    if (code === 2 || code === 5) continue;
    push(to, snap.edgeIndex, [], Math.max(1, remain), [snap.edgeIndex]);
    if (isDest) {
      return {
        ok: true,
        distanceMeters: Math.max(1, remain),
        edgeIndexes: [snap.edgeIndex],
        osmWayIds: [pack.osmWayIds[snap.edgeIndex]],
        unprovenStitches: 0
      };
    }
  }

  while (heap.length) {
    heap.sort((a, b) => a.cost - b.cost);
    const cur = heap.shift();
    const key = cur.node + ":" + cur.incomingEdge + ":" + activeKey(cur.restrictionActive);
    if (dist.get(key) !== cur.cost) continue;
    if (cur.cost > 1e9) break;
    const start = pack.nodeOffsets[cur.node];
    const end = pack.nodeOffsets[cur.node + 1];
    for (let i = start; i < end; i += 1) {
      const to = pack.edgeTargets[i];
      const ei = pack.edgeUndirectedIndex[i];
      if (ei === cur.incomingEdge) continue;
      const from = cur.node;
      const code = accessCode(pack, ei, from, to);
      const isDest = destEdges.has(ei);
      if (code === 2 || code === 5) continue;
      if (code === 1 && !options.allowUnknown) continue;
      if (!endpointOk(code, isDest, options.endEndpointKind || null) && code !== 1) continue;
      const restriction = advanceRestrictionState(
        restrictionIndex,
        cur.restrictionActive,
        cur.incomingEdge,
        ei,
        cur.node
      );
      if (!restriction.allowed) continue;
      const nextCost = cur.cost + pack.edgeMeters[ei];
      const pathEdges = cur.pathEdges.concat([ei]);
      if (isDest) {
        const ways = pathEdges.map((id) => pack.osmWayIds[id]);
        return {
          ok: true,
          distanceMeters: nextCost,
          edgeIndexes: pathEdges,
          osmWayIds: ways,
          unprovenStitches: 0
        };
      }
      push(to, ei, restriction.active, nextCost, pathEdges);
    }
  }
  return { ok: false, reason: "no_route", unprovenStitches: 0 };
}

module.exports = { findPathV4, haversineMeters };

File: scripts/pack-fabric/routing/kotlin/src/FindPathV4.kt
package dirt.routing.v4

data class LatLon(val lat: Double, val lon: Double)

data class PathV4Result(
    val ok: Boolean,
    val distanceMeters: Double = 0.0,
    val edgeIndexes: List<Int> = emptyList(),
    val osmWayIds: List<Long> = emptyList(),
    val reason: String? = null
)

object FindPathV4 {
    fun findPath(
        pack: GraphV4Pack,
        origin: LatLon,
        dest: LatLon,
        startHeadingDeg: Double? = null,
        intentBearingDeg: Double? = null,
        allowUnknown: Boolean = false,
        maxMeters: Double? = null,
        zoom: Double? = null
    ): PathV4Result {
        val startDetailed = LegalSnap.legalSnapDetailed(
            pack,
            origin,
            startHeadingDeg,
            intentBearingDeg,
            maxMeters,
            zoom,
            allowUnknown
        )
        val endIntent = intentBearingDeg?.let { (it + 180) % 360 }
        val endDetailed = LegalSnap.legalSnapDetailed(
            pack,
            dest,
            null,
            endIntent,
            maxMeters,
            zoom,
            allowUnknown
        )
        val picked = LegalSnap.selectConnectedSnapPair(
            pack,
            startDetailed.candidates,
            endDetailed.candidates,
            allowUnknown
        )
        if (!picked.ok || picked.start == null || picked.end == null) {
            return PathV4Result(ok = false, reason = picked.reason ?: "no_snap")
        }
        val startSnaps = listOf(picked.start)
        val endSnaps = listOf(picked.end)
        val destEdges = endSnaps.map { it.edgeIndex }.toSet()
        data class State(val node: Int, val incomingEdge: Int, val cost: Double, val pathEdges: List<Int>)
        val dist = HashMap<String, Double>()
        val heap = ArrayList<State>()
        fun key(node: Int, incoming: Int) = "$node:$incoming"
        fun push(node: Int, incomingEdge: Int, cost: Double, pathEdges: List<Int>) {
            val k = key(node, incomingEdge)
            val old = dist[k]
            if (old != null && old <= cost) return
            dist[k] = cost
            heap.add(State(node, incomingEdge, cost, pathEdges))
        }

        for (snap in startSnaps) {
            val from = if (snap.forward) pack.edgeFrom[snap.edgeIndex] else pack.edgeTo[snap.edgeIndex]
            val to = if (snap.forward) pack.edgeTo[snap.edgeIndex] else pack.edgeFrom[snap.edgeIndex]
            val remain = if (snap.forward) {
                (1 - snap.fraction) * pack.edgeMeters[snap.edgeIndex]
            } else {
                snap.fraction * pack.edgeMeters[snap.edgeIndex]
            }
            val code = pack.accessCode(snap.edgeIndex, from, to)
            val isDest = destEdges.contains(snap.edgeIndex)
            if (code != 0 && !(code == 3 || code == 4) && code != 1) {
                if (code == 2 || code == 5) continue
            }
            if (code == 2 || code == 5) continue
            if ((code == 3 || code == 4) && !isDest) continue
            push(to, snap.edgeIndex, maxOf(1.0, remain.toDouble()), listOf(snap.edgeIndex))
            if (isDest) {
                return PathV4Result(
                    ok = true,
                    distanceMeters = maxOf(1.0, remain.toDouble()),
                    edgeIndexes = listOf(snap.edgeIndex),
                    osmWayIds = listOf(pack.osmWayIds[snap.edgeIndex])
                )
            }
        }

        while (heap.isNotEmpty()) {
            heap.sortBy { it.cost }
            val cur = heap.removeAt(0)
            val k = key(cur.node, cur.incomingEdge)
            if (dist[k] != cur.cost) continue
            if (cur.cost > 1e9) break
            val start = pack.nodeOffsets[cur.node]
            val end = pack.nodeOffsets[cur.node + 1]
            for (i in start until end) {
                val to = pack.edgeTargets[i]
                val ei = pack.edgeUndirectedIndex[i]
                if (ei == cur.incomingEdge) continue
                val code = pack.accessCode(ei, cur.node, to)
                val isDest = destEdges.contains(ei)
                if (pack.hopIllegal(
                        ei,
                        cur.node,
                        to,
                        startEi = cur.pathEdges.firstOrNull() ?: -1,
                        endEi = if (isDest) ei else -1,
                        incomingEi = cur.incomingEdge
                    )
                ) {
                    continue
                }
                if (code == 1 && !allowUnknown) continue
                val nextCost = cur.cost + pack.edgeMeters[ei]
                val pathEdges = cur.pathEdges + ei
                if (isDest) {
                    return PathV4Result(
                        ok = true,
                        distanceMeters = nextCost,
                        edgeIndexes = pathEdges,
                        osmWayIds = pathEdges.map { pack.osmWayIds[it] }
                    )
                }
                push(to, ei, nextCost, pathEdges)
            }
        }
        return PathV4Result(ok = false, reason = "no_route")
    }
}
